### PR TITLE
rocksdb_replicator: reset upstream when fetching from a non-leader that has no updates

### PR DESCRIPTION
**Pull request by jaricftw**
We see in production where writes can timeout after 2-ack mode is enabled, which usually is caused by followers talking to the wrong upstream (typically a follower), hence never ACKing the write. See this internal [doc](https://docs.google.com/document/d/1xMX3Vkf3fne-usSowSaW9L2deKDDuMEGHLg1IXkxNTo/edit#heading=h.pik5pkd0ubqe) for detailed investigation and plans on fixing those issues. 

This PR serves as the initial step to mitigate the issue: trigger a upstream reset when both of the two conditions below are met:
- no updates from upstream, for a consecutive number of times (controlled by a GFLAG, in case it's transient)
- upstream informs the follower that the upstream is NOT a leader (avoids unnecessary reset when there is no traffic). We introduced the `ReplicaRole` in https://github.com/pinterest/rocksplicator/pull/592

This new behavior is disabled by default (via a GFLAG `reset_upstream_on_empty_updates_from_non_leader`) for a safe rollout, but can be reversed once this feature goes out well in production.

Added unit tests for the two edge cases we see in prod (followers setting each other as upstream; and follower setting itself as upstream).

### CODE REVIEW

File: rocksdb_replicator/rocksdb_replicator.h [hunk 1]
```diff
@@ -136,7 +136,9 @@ class RocksDBReplicator {
     std::shared_ptr<replicator::DbWrapper> db_wrapper_;
     folly::Executor* const executor_;
     const DBRole role_;
+    const ReplicaRole role_thrift_;
```
Comment (kangnanli) [inline rocksdb_replicator/rocksdb_replicator.h:139]: Why do we need two *roles*? Won't DBRole is actually the replication role?

File: rocksdb_replicator/replicated_db.cpp [hunk 2]
```diff
@@ -317,7 +331,23 @@ void RocksDBReplicator::ReplicatedDB::pullFromUpstream() {
...
           if (!response.updates.empty()) {
+            db->pullFromUpstreamNoUpdates_ = 0;
             db->cond_var_.notifyAll();
+          } else {
+            incCounter(kReplicatorPullRequestsNoUpdates, 1, db->db_name_);
+            // no updates consecutively, and the upstream says it's NOT a leader.
+            // Therefore we reset upstream. 
+            db->pullFromUpstreamNoUpdates_++;
+            if (response.role != ReplicaRole::LEADER
+                && FLAGS_reset_upstream_on_empty_updates_from_non_leader
```
Comment (kangnanli) [inline rocksdb_replicator/replicated_db.cpp:342]: could you elaborate why we need `reset_upstream_on_empty_updates_from_non_leader`? 

when replicated from a nonleader && empty response, shouldn't we reset upstream immediately?

File: rocksdb_replicator/rocksdb_replicator.h (hunk 1 above)
Comment (jaricftw) [reply rocksdb_replicator/rocksdb_replicator.h:139]: This is an optimization: we shouldn't need to pass string over the wire on every replication response which can be expensive

File: rocksdb_replicator/replicated_db.cpp (hunk 2 above)
Comment (jaricftw) [reply rocksdb_replicator/replicated_db.cpp:342]: This FLAG is for a safe rollout so we don't introduce this change immediately when upgrading each cluster. Once we verified this works for a few clusters, perhaps we can enable this feature by default

> when replicated from a nonleader && empty response, shouldn't we reset upstream immediately?
yep, that's the goal state. Today we want to be a little conservative to avoid overloading Helix/ZK. The plan is to reset immediately (PRs will be coming to achieve this goal, and if you are interested, check the doc I linked in the description).

File: rocksdb_replicator/rocksdb_replicator.h (hunk 1 above)
Comment (kangnanli) [reply rocksdb_replicator/rocksdb_replicator.h:139]: `DBRole` is *enum*, right? not a string: https://github.com/pinterest/rocksplicator/blob/master/rocksdb_replicator/rocksdb_replicator.h#L69

Comment (jaricftw) [reply rocksdb_replicator/rocksdb_replicator.h:139]: Sorry, should have made it clearer: we cannot pass `DBRole` directly into the thrift replication response, so we would need to define a thrift enum instead of the `DBRole` enum as you pointed above. 

I thought about using the thrift enum everywhere and get rid of the existing `DBRole` enum, but that will cause a large blast radius of refactor: https://github.com/pinterest/rocksplicator/search?p=3&q=DBRole. Something I can follow up. 

There is also another difference I forgot to callout: `DBRole` has `NOOP` as well which is a validate replication role (e.g. in [BOOTSTRAP](https://github.com/pinterest/rocksplicator/blob/63c9c7d738dc868e59c552a43e2b305314092b0c/cluster_management/src/main/java/com/pinterest/rocksplicator/BootstrapStateModelFactory.java#L129) mode I think), but for the replication flow, only LEADER/FOLLOWER is concerned, so these two enums are not exactly the same.

Comment (kangnanli) [reply rocksdb_replicator/rocksdb_replicator.h:139]: I think we need unify them, it will cause confusion. 

There is really only 1 role for the db, which is defined by *Replication*. 
The *Noop* just tell the replicationDB that it does not replicate from anyone. https://github.com/pinterest/rocksplicator/blob/63c9c7d738dc868e59c552a43e2b305314092b0c/rocksdb_admin%2Fadmin_handler.cpp#L616-L617

Comment (kangnanli) [reply rocksdb_replicator/rocksdb_replicator.h:139]: On top of the cpp, at the cluster management layer, we have another *role*, which Helix partition role, such as Helix partition Leader, Folllower, etct, which defined by statemodel.

File: rocksdb_replicator/thrift/replicator.thrift [hunk 3]
```diff
@@ -53,10 +53,18 @@ struct Update {
   3: optional i64 seq_no,
 }
 
+enum ReplicaRole {
+  OTHER = 0,
```
Comment (kangnanli) [inline rocksdb_replicator/thrift/replicator.thrift:57]: I think, this should be *NoOP*.

File: rocksdb_replicator/thrift/replicator.thrift [hunk 4]
```diff
@@ -53,10 +53,18 @@ struct Update {
   3: optional i64 seq_no,
 }
 
+enum ReplicaRole {
```
Comment (kangnanli) [inline rocksdb_replicator/thrift/replicator.thrift:56]: maybe considering renaming to *ReplicatedRole* to make it explicit.

File: rocksdb_replicator/rocksdb_replicator.h (hunk 1 above)
Comment (kangnanli) [reply rocksdb_replicator/rocksdb_replicator.h:139]: Should we consider other names than `role_thrift_`? Maybe, `replica_role_` if the class is `ReplicaRole`

Comment (jaricftw) [reply rocksdb_replicator/rocksdb_replicator.h:139]: Makes sense, In that case, does my above proposal works for you?
> I thought about using the thrift enum everywhere and get rid of the existing DBRole enum, but that will cause a large blast radius of refactor: https://github.com/pinterest/rocksplicator/search?p=3&q=DBRole. Something I can follow up.

Or do you prefer we do that first before this change. I am okay either way

Comment (jaricftw) [reply rocksdb_replicator/rocksdb_replicator.h:139]: I will try put up a PR anyway

Comment (jaricftw) [reply rocksdb_replicator/rocksdb_replicator.h:139]: See https://github.com/pinterest/rocksplicator/pull/592, can you review that first

**Comment by jaricftw** [comment]
@kangnanli or @rajathprasad this is ready for another look

File: rocksdb_replicator/replicated_db.cpp [hunk 5]
```diff
@@ -317,7 +331,23 @@ void RocksDBReplicator::ReplicatedDB::pullFromUpstream() {
...
 
           if (!response.updates.empty()) {
+            db->pullFromUpstreamNoUpdates_ = 0;
             db->cond_var_.notifyAll();
+          } else {
+            incCounter(kReplicatorPullRequestsNoUpdates, 1, db->db_name_);
+            // no updates consecutively, and the upstream says it's NOT a leader.
+            // Therefore we reset upstream.
+            db->pullFromUpstreamNoUpdates_++;
+            if (response.role != ReplicaRole::LEADER
```
Comment (kangnanli) [inline rocksdb_replicator/replicated_db.cpp:341]: Do you want to add a counter to record pull from Non-leader for replication?

File: rocksdb_replicator/tests/rocksdb_replicator_test.cpp [hunk 6]
```diff
@@ -348,6 +361,126 @@ TEST(RocksDBReplicatorTest, 1_master_2_slaves_chain) {
   EXPECT_EQ(db_slave_2->GetLatestSequenceNumber(), 2 * n_keys);
 }
 
+TEST(RocksDBReplicatorTest, 1_master_1_slave_upstream_itself) {
```
Comment (kangnanli) [inline rocksdb_replicator/tests/rocksdb_replicator_test.cpp:364]: I think, for the test name, we still keep the camel case, not `1_master_1_slave_upstream_itself`

File: rocksdb_replicator/tests/rocksdb_replicator_test.cpp [hunk 7]
```diff
@@ -348,6 +361,126 @@ TEST(RocksDBReplicatorTest, 1_master_2_slaves_chain) {
...
+    WriteBatch updates;
+    auto str = to_string(i);
+    updates.Put(str + "key", str + "value");
+    updates.Put(str + "key2", str + "value2");
+    EXPECT_EQ(master.replicator_->write("shard1", options, &updates),
+              ReturnCode::OK);
+    EXPECT_EQ(db_master->GetLatestSequenceNumber(), i * 2 + 2);
+  }
+
+  // expect follower reset upstream to be triggered
```
Comment (kangnanli) [inline rocksdb_replicator/tests/rocksdb_replicator_test.cpp:407]: when the follower trigger the `resetUpstream`, it seem like the `resetUpstream` will request zookeeper?
Now, at this unit test, there is no zk setup ,right? Then, what is the side effect of that `resetUpstream` call?

File: rocksdb_replicator/tests/rocksdb_replicator_test.cpp (hunk 6 above)
Comment (kangnanli) [inline rocksdb_replicator/tests/rocksdb_replicator_test.cpp:364]: Could you add a few lines of comments to explain:
- what is the purpose of test, and how would you plan to test it?
( I can get these info by reading the code, but, I think it might beneficial to briefly explain it)

**Review by kangnanli — APPROVED** [review]
I don't have any block reviews. 
Fix before land

File: rocksdb_replicator/tests/rocksdb_replicator_test.cpp (hunk 6 above)
Comment (jaricftw) [reply rocksdb_replicator/tests/rocksdb_replicator_test.cpp:364]: I am using this style for consistency with all existing tests in this file. If we decide to use CamelCase, then we can change it for all

Comment (jaricftw) [reply rocksdb_replicator/tests/rocksdb_replicator_test.cpp:364]: added

File: rocksdb_replicator/replicated_db.cpp (hunk 5 above)
Comment (jaricftw) [reply rocksdb_replicator/replicated_db.cpp:341]: good idea, added `kReplicatorPullFromNonLeader`

File: rocksdb_replicator/tests/rocksdb_replicator_test.cpp (hunk 7 above)
Comment (jaricftw) [reply rocksdb_replicator/tests/rocksdb_replicator_test.cpp:407]: yeah, it'll just be a no-op: https://github.com/pinterest/rocksplicator/blob/bec3645b49c996e3321faac5ab52b9fe04df1a51/rocksdb_replicator/replicated_db.cpp#L231, I have a comment/test below to verify the `resetUpstreamAttempts_` is incremented, but no sequence number change (because reset won't succeed due to this)

**Comment by jaricftw** [comment]
Landing, this won't have any production impact, unless explicitly enabled via the GFLAG.
